### PR TITLE
Prepare to use subprograms (emit BTF)

### PR DESCRIPTION
**Pull request by viktormalik**
This prepares bpftrace for the use of subprograms necessary for #2516.

There are multiple steps that need to be done:
1. debug info is generated for all function definitions (LLVM will convert it to `.BTF`/`.BTF.ext` sections)
2. subprograms (located in the `.text` section) are appended to the main program (probes)
3. subprograms are relocated by virtue of `R_BPF_64_64` entries
4. BTF is passed to libbpf upon program loading

Note that this is just a minimal required effort to support subprograms and there are several manual hacks for dealing with points 3 and 4. Implementing them properly would require changes to the overall bpftrace architecture and the proposed approach should be able to handle most of the subprogram support that we need at this moment.

This also includes some refactoring of the way BPF bytecode is stored inside bpftrace.

This PR is based on #2321 by separating the relevant commits from there.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

### CODE REVIEW

**Comment by danobi** [comment]
Will try to review this weekend

**Comment by ajor** [comment]
I've had a quick skim but there's a lot in here that I don't understand. I'd like to give it a thorough review myself as well before merging, but that'll most likely be in a couple of weeks as I'm off on holiday now. Sorry for the delay

**Review by danobi — COMMENTED** [review]
only looked at first 3 commits so far

File: src/ast/passes/codegen_llvm.cpp [hunk 1]
```diff
@@ -3296,6 +3331,88 @@ void CodegenLLVM::generateWatchpointSetupProbe(
   createRet();
 }
 
+std::string CodegenLLVM::sanitizeProgramName(std::string const &name)
```
[the remark this answers is not in the document]
Comment (viktormalik) [reply src/ast/passes/codegen_llvm.cpp]: Good point, I added a comment.

I also realized that we use sanitisation of BPF program names from other places (`attached_probe.cpp`), so I moved the function to `utils`. The problem is that `utils` doesn't link against LLVM so I had to change `llvm::SHA1` for `std::hash` but that's IMHO ok for this usecase.

File: src/bpfprogram.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include <optional>
+#include <string>
+#include <tuple>
+
+namespace bpftrace {
+
+class BpfProgram
+{
+public:
+  static std::optional<BpfProgram> CreateFromBytecode(BpfBytecode &bytecode,
```
[the remark this answers is not in the document]
Comment (viktormalik) [reply src/bpfprogram.h]: It's not, marked `const`.

Note: the same applies to `MapManager` added in a later commit. The problem is that `MapManager` is our own container class and it doesn't implement any `const` lookup methods. I don't think that this usecase is worth the trouble of adding them, but feel free to disagree with me.

**Comment by viktormalik** [comment]
Posted first part of changes in #2791.

**Comment by viktormalik** [comment]
This is now fully merged piece-wise by #2791, #2804, and #2840.

Thanks everyone for the reviews and a lot of feedback, this is a very important infrastructure!

**Pull request closed without merging**